File: solutions/cpp/two-bucket/1/two_bucket.cpp
#include "two_bucket.h"

#include <stdexcept>

namespace two_bucket
{

    unsigned int gcdx(unsigned int a, unsigned int b, int &u, int &v){
    if (b==0) {
        u = 1;
        v = 0;
        return a;
    }
    int u1, v1;
    unsigned int d = gcdx(b, a%b, u1, v1);
    u = v1;
    v = u1 - v1*(a/b);
    return d;
}


    measure_result measure_one(int bucket1_capacity,
                        int bucket2_capacity,
                        int target_volume){
        if (target_volume==bucket1_capacity)
            return measure_result{1,bucket_id::one,0};
        else if (target_volume==bucket2_capacity)
            return measure_result{2,bucket_id::two,bucket1_capacity};
        int u,v;
        int d = gcdx(bucket1_capacity, bucket2_capacity, u, v);
        if(target_volume % d !=0)
            throw std::invalid_argument("No possible.");
        u *= target_volume/d;
        v *= target_volume/d;
        if (u<0) {
            int t = -u*d/bucket2_capacity + 1;
            u += (bucket2_capacity/d)*t;
            v -= (bucket1_capacity/d)*t;
        }
        if (bucket1_capacity>=target_volume)
            return measure_result{2*(u-v) -2, bucket_id::one, bucket2_capacity};
        else
            return measure_result{2*(u-v), bucket_id::two, 0};
    }


    measure_result measure(int bucket1_capacity,
                        int bucket2_capacity,
                        int target_volume, bucket_id start_bucket){
        if(!target_volume)
             return measure_result{0,bucket_id::one,0};
        else if (target_volume>bucket1_capacity &&
             target_volume>bucket2_capacity)
             throw std::invalid_argument("Not possible.");
        if(start_bucket == bucket_id::one)
            return measure_one(bucket1_capacity, bucket2_capacity, target_volume);
        else {
            measure_result res =
                measure_one(bucket2_capacity, bucket1_capacity, target_volume);
            res.goal_bucket = 
                (res.goal_bucket == bucket_id::one) ? bucket_id::two : bucket_id::one;
            return res;
        }
    }
    
    
}  // namespace two_bucket

File: solutions/dart/two-bucket/1/lib/two_bucket.dart
typedef Result = ({int moves, String goalBucket, int otherBucket});

class TwoBucket {

  int capBucketOne = 0;
  int capBucketTwo = 0;
  int goalVol = 0;
  String start = '';
  
  TwoBucket({int? bucketOne, int? bucketTwo, int? goal, String? startBucket}){
    this.capBucketOne = bucketOne ?? 0;
    this.capBucketTwo = bucketTwo ?? 0;
    this.goalVol = goal ?? 0;
    this.start = startBucket ?? "one";
  }
  
  List<int> gcdx(int a, int b){
    var res = List<int>.filled(3,0);
    if(b==0){
      res[0] = a; res[1] = 1; res[2] = 0;
    }
    else{
      res = gcdx(b, a%b);
      var u1 = res[2];
      res[2] = res[1] - res[2]*(a~/b);
      res[1] = u1;
    }
    return res;      
  }

  Result measureOne(int capBucket1, int capBucket2, int targetVol){
    if (targetVol == capBucket1)
      return (moves: 1, goalBucket: "one", otherBucket: 0);
    else if (targetVol == capBucket2)
      return (moves:2, goalBucket:"two", otherBucket:capBucket1);
    var gcd = gcdx(capBucket1, capBucket2);
    if(targetVol % gcd[0] !=0 )
      throw ArgumentError('impossible goal');
    int u = gcd[1] * (targetVol ~/ gcd[0]);
    int v = gcd[2] * (targetVol ~/ gcd[0]);
    if(u<0){
       int t = -u*gcd[0]~/capBucket2 + 1;
        u += (capBucket2~/gcd[0])*t;
        v -= (capBucket1~/gcd[0])*t;
    }
    if(capBucket1 > targetVol)
      return (moves: 2*(u-v)-2, goalBucket: "one", otherBucket: capBucket2);
    else
      return (moves: 2*(u-v), goalBucket: "two", otherBucket: 0);    
  }

  Result measure(){
    if(goalVol == 0)
      return (moves: 0, goalBucket: "one", otherBucket: 0);
    else if(goalVol > capBucketOne
            && goalVol > capBucketTwo)
      throw ArgumentError('goal larger than both buckets is impossible');
    if (start == "one")
      return this.measureOne(capBucketOne, capBucketTwo, goalVol);
    else {
      Result res = this.measureOne(capBucketTwo, capBucketOne, goalVol);
      if (res.goalBucket == 'one') 
        return (moves: res.moves, goalBucket: 'two', otherBucket: res.otherBucket);
      else 
        return (moves: res.moves, goalBucket: 'one', otherBucket: res.otherBucket);
    }
  }
  
}
